File: src/main/java/net/unaussprechlich/warlordsplus/hud/elements/HudElementTotalKills.java
package net.unaussprechlich.warlordsplus.hud.elements;

import net.minecraft.client.Minecraft;
import net.minecraft.util.EnumChatFormatting;
import net.minecraftforge.client.event.ClientChatReceivedEvent;
import net.unaussprechlich.warlordsplus.WarlordsPlus;
import net.unaussprechlich.warlordsplus.hud.AbstractHudElement;
import net.unaussprechlich.warlordsplus.module.modules.GameStateManager;
import net.unaussprechlich.warlordsplus.util.consumers.IChatConsumer;
import net.unaussprechlich.warlordsplus.util.consumers.IUpdateConsumer;

public class HudElementTotalKills extends AbstractHudElement implements IUpdateConsumer, IChatConsumer {

    TeamEnum team = TeamEnum.NONE;
    private int redKills = 0;
    private int blueKills = 0;


    enum TeamEnum {
        BLUE, RED, NONE
    }

    @Override
    public String[] getRenderString() {
        return new String[]{EnumChatFormatting.RED + "Red Kills: " + redKills, EnumChatFormatting.BLUE + "Blue Kills: " + blueKills};
    }

    @Override
    public void update() {
        if (Minecraft.getMinecraft().thePlayer != null) {
            if (Minecraft.getMinecraft().thePlayer.getDisplayName().getFormattedText().contains("\u00A7c")) {
                team = TeamEnum.RED;
            } else if (Minecraft.getMinecraft().thePlayer.getDisplayName().getFormattedText().contains("\u00A79")) {
                team = TeamEnum.BLUE;
            } else {
                team = TeamEnum.NONE;
            }
        }
    }


    @Override
    public void onChat(ClientChatReceivedEvent e) {
        String message = e.message.getFormattedText();

        if (message.contains("The gates will fall")) {
            redKills = 0;
            blueKills = 0;
        }

        if (message.contains("was killed by")) {
            String after = message.substring(message.indexOf("was killed by"));
            if ((after.contains("\u00A7c"))) {
                redKills++;
            } else
                blueKills++;
        }

        if (message.contains("You killed")) {
            if (team == TeamEnum.RED)
                redKills++;
            else if (team == TeamEnum.BLUE)
                blueKills++;
        }

        if (message.contains("You were killed")) {
            if (team == TeamEnum.RED)
                blueKills++;
            else if (team == TeamEnum.BLUE)
                redKills++;
        }

        if (message.contains("You assisted in killing")) {
            if (team == TeamEnum.RED)
                redKills++;
            else if (team == TeamEnum.BLUE)
                blueKills++;
        }
    }

    @Override
    public boolean isVisible() {
        return GameStateManager.INSTANCE.isIngame();
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: src/main/java/net/unaussprechlich/warlordsplus/hud/elements/HudElementTotalKills.kt
package net.unaussprechlich.warlordsplus.hud.elements

import net.minecraft.util.EnumChatFormatting
import net.minecraftforge.client.event.ClientChatReceivedEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import net.unaussprechlich.eventbus.EventBus
import net.unaussprechlich.warlordsplus.OtherPlayers
import net.unaussprechlich.warlordsplus.config.CCategory
import net.unaussprechlich.warlordsplus.config.ConfigPropertyBoolean
import net.unaussprechlich.warlordsplus.hud.AbstractHudElement
import net.unaussprechlich.warlordsplus.module.modules.GameStateManager
import net.unaussprechlich.warlordsplus.module.modules.GameStateManager.isIngame
import net.unaussprechlich.warlordsplus.module.modules.KillEvent
import net.unaussprechlich.warlordsplus.module.modules.ResetEvent
import net.unaussprechlich.warlordsplus.util.TeamEnum
import net.unaussprechlich.warlordsplus.util.removeFormatting
import java.util.*

object HudElementTotalKills : AbstractHudElement() {

    var blueKills: Int = 0
    var redKills: Int = 0
    private var numberOfCapsBlue: Int = 0
    private var numberOfCapsRed: Int = 0

    init {
        EventBus.register<ResetEvent> {
            redKills = 0
            blueKills = 0
            numberOfCapsRed = 0
            numberOfCapsBlue = 0
        }

        EventBus.register<KillEvent> {
            if (GameStateManager.isDOM) {
                if (OtherPlayers.getPlayerForName(it.player)!!.team == TeamEnum.BLUE)
                    blueKills++
                else if (OtherPlayers.getPlayerForName(it.player)!!.team == TeamEnum.RED)
                    redKills++
            }

        }

        EventBus.register<ClientChatReceivedEvent> {
            if (GameStateManager.isCTF) {
                val message = it.message.unformattedText.removeFormatting()
                if (message.contains("captured the")) {
                    if (message.contains("RED")) {
                        numberOfCapsBlue++
                    } else {
                        numberOfCapsRed++
                    }
                }

            }
        }

        EventBus.register<TickEvent.ClientTickEvent> {
            if (GameStateManager.isCTF) {
                blueKills = (GameStateManager.bluePoints - numberOfCapsBlue * 250) / 5
                redKills = (GameStateManager.redPoints - numberOfCapsRed * 250) / 5
            } else if (GameStateManager.isTDM) {
                blueKills = GameStateManager.bluePoints / 15
                redKills = GameStateManager.redPoints / 15
            }
        }
    }


    override fun getRenderString(): Array<String> {
        val renderStrings = ArrayList<String>()

        if (showBlueKills)
            renderStrings.add(EnumChatFormatting.BLUE.toString() + "Blue Kills: " + blueKills)
        if (showRedKills)
            renderStrings.add(EnumChatFormatting.RED.toString() + "Red Kills: " + redKills)

        return renderStrings.toTypedArray()
    }

    override fun isVisible(): Boolean {
        return isIngame && !GameStateManager.isPvE
    }

    override fun isEnabled(): Boolean {
        return true
    }

    @ConfigPropertyBoolean(
        category = CCategory.HUD,
        id = "|| Total Kills Blue | Show",
        comment = "Enable or disable the Blue Kill counter",
        def = true
    )
    var showBlueKills = false

    @ConfigPropertyBoolean(
        category = CCategory.HUD,
        id = "|| Total Kills Red | Show",
        comment = "Enable or disable the Red Kill counter",
        def = true
    )
    var showRedKills = false


}